lib/extras/gain_map.cc

```diff
@@ -0,0 +1,216 @@
+  }
+
+  uint32_t icc_size_le = JXL_BSWAP32(map_bundle->alt_icc_size);
+  if (cursor + 4 <= output_buffer_size) {
+    memcpy(output_buffer + cursor, &icc_size_le, 4);
+    cursor += 4;
+  }
+
+  if (cursor + map_bundle->alt_icc_size <= output_buffer_size) {
+    memcpy(output_buffer + cursor, map_bundle->alt_icc,
+           map_bundle->alt_icc_size);
+    cursor += map_bundle->alt_icc_size;
+  }
+
+  if (cursor + map_bundle->gain_map_size <= output_buffer_size) {
+    memcpy(output_buffer + cursor, map_bundle->gain_map,
+           map_bundle->gain_map_size);
+    cursor += map_bundle->gain_map_size;
+  }
+
+  if (bytes_written != nullptr) *bytes_written = cursor;
+  return cursor == output_buffer_size ? JXL_TRUE : JXL_FALSE;
+}
+
+JXL_BOOL JxlGainMapReadBundle(JxlGainMapBundle* map_bundle,
+                              const uint8_t* input_buffer,
+                              const size_t input_buffer_size,
+                              size_t* bytes_read) {
+  if (map_bundle == nullptr || input_buffer == nullptr ||
+      input_buffer_size == 0) {
+    return JXL_FALSE;
+  }
+
+  size_t cursor = 0;
+  // Read the version byte
+  map_bundle->jhgm_version = input_buffer[cursor];
+  cursor += 1;
+
+  // Read and swap gain_map_metadata_size
+  uint16_t gain_map_metadata_size_le;
+  memcpy(&gain_map_metadata_size_le, input_buffer + cursor, 2);
+  uint16_t gain_map_metadata_size = JXL_BSWAP16(gain_map_metadata_size_le);
+  cursor += 2;
+  map_bundle->gain_map_metadata_size = gain_map_metadata_size;
+  map_bundle->gain_map_metadata = input_buffer + cursor;
+  cursor += gain_map_metadata_size;
+
+  // Read compressed_color_encoding_size
+  uint8_t compressed_color_encoding_size;
+  memcpy(&compressed_color_encoding_size, input_buffer + cursor, 1);
+  cursor += 1;
+
+  map_bundle->has_color_encoding = (0 < compressed_color_encoding_size);
+  if (map_bundle->has_color_encoding) {
+    // Decode color encoding
+    jxl::Span<const uint8_t> color_encoding_span(
+        input_buffer + cursor, compressed_color_encoding_size);
+    jxl::BitReader color_encoding_reader(color_encoding_span);
+    jxl::ColorEncoding internal_color_encoding;
+    if (!jxl::Bundle::Read(&color_encoding_reader, &internal_color_encoding)) {
+      return JXL_FALSE;
+    }
+    JXL_RETURN_IF_ERROR(color_encoding_reader.Close());
+    map_bundle->color_encoding = internal_color_encoding.ToExternal();
+  }
+  cursor += compressed_color_encoding_size;
+  // Read and swap compressed_icc_size
+  uint32_t compressed_icc_size_le;
+  memcpy(&compressed_icc_size_le, input_buffer + cursor, 4);
+  uint32_t compressed_icc_size = JXL_BSWAP32(compressed_icc_size_le);
+  cursor += 4;
+  map_bundle->alt_icc_size = compressed_icc_size;
+  map_bundle->alt_icc = input_buffer + cursor;
+  cursor += compressed_icc_size;
+  // Remaining bytes are gain map
+  map_bundle->gain_map_size = input_buffer_size - cursor;
+  if (input_buffer_size < cursor + map_bundle->gain_map_size) return JXL_FALSE;
```

Comment:
I'm pretty sure this condition does nothing ;-)

Reply:
now with the SafeAdd there is a possibilty of failure